File: competitive_programming/leetcode/06_javascript/14_longest_common_prefix.js
function longestCommonPrefix(strs) {
  if (strs.length === 0) return "";
  let prefix = strs[0];
  for (const word of strs.slice(1)) {
    while (!word.startsWith(prefix)) {
      prefix = prefix.slice(0, -1);
      if (!prefix) return "";
    }
  }
  return prefix;
}

if (require.main === module) {
  if (longestCommonPrefix(["flower", "flow", "flight"]) !== "fl") throw new Error("longest common prefix failed");
  if (longestCommonPrefix(["dog", "racecar", "car"]) !== "") throw new Error("longest common prefix failed");
  console.log("14 longest common prefix ok");
}

File: competitive_programming/leetcode/09_rust/14_longest_common_prefix.rs
fn longest_common_prefix(strs: &[String]) -> String {
    if strs.is_empty() { return String::new(); }
    let mut prefix = strs[0].clone();
    for word in &strs[1..] {
        while !word.starts_with(&prefix) {
            prefix.pop();
            if prefix.is_empty() { return String::new(); }
        }
    }
    prefix
}

fn main() {
    let words: Vec<String> = ["flower", "flow", "flight"].iter().map(|s| s.to_string()).collect();
    assert_eq!(longest_common_prefix(&words), "fl");
    let none: Vec<String> = ["dog", "racecar", "car"].iter().map(|s| s.to_string()).collect();
    assert_eq!(longest_common_prefix(&none), "");
    println!("14 longest common prefix ok");
}
